File: code/904. Fruit Into Baskets.py
class Solution:
    def totalFruit(self, tree: List[int]) -> int:
        count = dict()
        left, ans, cnt = 0, 0, 0
        for right in range(len(tree)):
            count[tree[right]] = count.get(tree[right], 0) + 1
            if count[tree[right]] == 1:
                cnt += 1
            while cnt > 2:
                count[tree[left]] -= 1
                if count[tree[left]] == 0:
                    cnt -= 1
                left += 1
            ans = max(right - left + 1, ans)
        return ans

File: code/904. Fruit Into Baskets.cpp
class Solution {
public:
    int totalFruit(vector<int>& tree) {
        int cnt = 0;
        int left = 0;
        int ans = 0;
        vector<int> count(tree.size(), 0);
        for (int right = 0; right < tree.size(); right++) {
            count[tree[right]]++;
            if (count[tree[right]] == 1) cnt++;
            while (cnt > 2) {
                count[tree[left]]--;
                if (count[tree[left]] == 0) cnt--;
                left++;
            }
            ans = max(ans, right - left + 1);
        }
        return ans;
    }
};